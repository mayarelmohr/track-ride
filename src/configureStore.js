import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import { persistStore, persistReducer } from "redux-persist";
import localForage from "localforage";
import rootReducer from "./reducers";
import { OrderedMap } from "immutable";
import { createTransform } from "redux-persist";

const SetTransform = createTransform(
  (inboundState) => {
    return { ...inboundState, stations: inboundState.stations.toArray() };
  },
  (outboundState) => {
    return {
      ...outboundState,
      stations: new OrderedMap(outboundState.stations),
    };
  },
  { whitelist: ["trip"] }
);

/* eslint-disable no-underscore-dangle */
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
/* eslint-enable */
const persistConfig = {
  key: "0.1.1",
  storage: localForage,
  whitelist: ["trip"],
  transforms: [SetTransform],
};
const persistedReducer = persistReducer(persistConfig, rootReducer);

const configureStore = (initialState = {}) => {
  const store = createStore(
    persistedReducer,
    initialState,
    composeEnhancers(applyMiddleware(thunk))
  );
  const persistor = persistStore(store);
  return { store, persistor };
};

export const { store, persistor } = configureStore(window.REDUX_STATE || {});
