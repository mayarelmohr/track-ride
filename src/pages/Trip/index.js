import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import GoogleMap from "../../components/Map";
import AddPassengerFrom from "../../components/AddPassengerForm";
import Bookings from "../../components/Bookings";
import TripInformation from "../../components/TripInformation";
import { csv } from "d3";
import {
  setStations,
  setTripStartTime,
  setTripState,
  updateBookings,
  updateCurrentStation,
  setBookings,
  setTripTime,
} from "../../reducers/trip";
import Button from "../../components/Common/Button";
import Loading from "../../components/Common/Loading";
import {
  mapDirectionsToPath,
  getDistance,
  getStartPoint,
} from "../../selectors/stations";
import styles from "./styles.module.css";
import { TRIP_TIME, TRIP_STATE } from "../../helpers/constants";

const Trip = (props) => {
  const [isLoading, setLoading] = useState(true);
  const [isBookRideFormVisible, setBookRideFormVisibility] = useState(false);
  const [markerLocation, setMarkerLocation] = useState({ lat: 0, lng: 0 });
  const requestAnimationFrameRef = React.useRef();
  const rideTimeoutRef = React.useRef();
  const readCSV = async () => {
    const stationsData = await csv("./data/route.csv");
    props.setStationsAction(stationsData);
    const usersData = await csv("./data/users.csv");
    props.setBookingsAction(usersData);
    props.setTripTimeAction(TRIP_TIME);
  };
  const {
    bookingsList,
    distance,
    tripTime,
    paths,
    startTime,
    tripState,
    isDataReady,
  } = props;

  useEffect(() => {
    if (tripState === TRIP_STATE.TRACKING) {
      rideTimeoutRef.current = setTimeout(() => {
        startRide(paths, TRIP_TIME, startTime);
      }, 300);
    }
    if (!isDataReady) {
      readCSV();
      setLoading(false);
    } else {
      setLoading(false);
    }
    return () => {
      cancelAnimationFrame(requestAnimationFrameRef.current);
      clearTimeout(rideTimeoutRef.current);
    };
  }, []);

  const startRide = (paths, duration, start) => {
    const move = (previousStationIndex = 0) => {
      const currentTime = Date.now();
      /* calculates time elapsed time since the last requested frame */
      const elapsedTime = currentTime - start;
      /* calculates the expected distance that car should have moved during this elapsed time */
      const progress = elapsedTime / duration;
      /* the next lat and lng based on the car should have made */
      const nextLatLngIndex = Math.floor(progress * paths.length);
      if (elapsedTime < duration && nextLatLngIndex < paths.length) {
        const currentStopPoint = paths[nextLatLngIndex];
        const { lat, lng, stationIndex } = currentStopPoint;
        setMarkerLocation({ lat, lng });
        if (previousStationIndex !== stationIndex) {
          props.updateCurrentStationAction(stationIndex);
          props.updateBookingsAction(stationIndex);
        }
        requestAnimationFrameRef.current = requestAnimationFrame(() => {
          move(stationIndex);
        });
      } else {
        cancelAnimationFrame(requestAnimationFrameRef.current);
        props.setTripStateAction(TRIP_STATE.FINISHED);
        props.history.push("/statistics");
      }
    };
    requestAnimationFrameRef.current = requestAnimationFrame(move);
  };

  if (isLoading) {
    return <Loading />;
  }
  return (
    <div>
      <GoogleMap markerLocation={markerLocation} data-testid="map" />
      <div className={styles.container}>
        <div className={styles.buttons}>
          <Button
            text="Start ride"
            data-testid="start-ride-button"
            type="button"
            disabled={tripState === TRIP_STATE.TRACKING}
            action={() => {
              props.setTripStateAction(TRIP_STATE.TRACKING);
              const start = Date.now();
              props.updateCurrentStationAction(0);
              props.setTripStartTimeAction(start);
              startRide(paths, TRIP_TIME, start);
            }}
          />
          <Button
            text="Book ride"
            data-testid="book-ride-button"
            type="button"
            disabled={tripState !== TRIP_STATE.IDLE}
            action={() => {
              setBookRideFormVisibility(true);
            }}
          />
        </div>
        <TripInformation distance={distance} time={tripTime} />
        <Bookings bookingsList={bookingsList} />
        <AddPassengerFrom
          isModalVisible={isBookRideFormVisible}
          closeModal={() => setBookRideFormVisibility(false)}
          contentLabel="Add credit card"
        />
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const {
    directions,
    tripTime,
    tripState,
    isDataReady,
    startTime,
    currentStation,
    stations,
  } = state.trip;
  const paths = mapDirectionsToPath(directions);
  const distance = getDistance(directions);
  const bookings =
    stations.get(currentStation)?.bookings ||
    getStartPoint(stations)?.bookings ||
    [];

  return {
    startTime,
    isDataReady,
    distance,
    tripTime,
    tripState,
    paths,
    bookingsList: bookings,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setStationsAction: (data) => dispatch(setStations(data)),
    setBookingsAction: (data) => dispatch(setBookings(data)),
    updateBookingsAction: (index) => dispatch(updateBookings(index)),
    setTripTimeAction: (time) => dispatch(setTripTime(time)),
    setTripStateAction: (state) => dispatch(setTripState(state)),
    setTripStartTimeAction: (time) => dispatch(setTripStartTime(time)),
    updateCurrentStationAction: (index) =>
      dispatch(updateCurrentStation(index)),
  };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Trip));
