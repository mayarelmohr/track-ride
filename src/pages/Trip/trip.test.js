// counter.test.js
import React from "react";
import "@testing-library/jest-dom/extend-expect";
import Trip from "./index";
import { Provider } from "react-redux";
import { render, fireEvent, cleanup } from "@testing-library/react";
import { store } from "../../configureStore";

afterEach(cleanup);

test("can render with redux with defaults", () => {
  const { debug, getByText } = render(
    <Provider store={store}>
      <Trip />
    </Provider>
  );
  const button = getByText("Start ride");
  fireEvent.click(button);
  debug();
});

test("snapshot trip 1", () => {
  const tripElement = render(
    <Provider store={store}>
      <Trip />
    </Provider>
  );
  expect(tripElement).toMatchSnapshot();
});
