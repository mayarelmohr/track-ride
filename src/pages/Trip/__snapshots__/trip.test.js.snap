// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot trip 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          style="height: 100%;"
        />
        <div
          class="container"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
              name="Start ride"
              type="button"
            >
              Start ride
            </button>
            <button
              class="button"
              name="Book ride"
              type="button"
            >
              Book ride
            </button>
          </div>
          <section>
            <h2>
              Trip Information
            </h2>
            <p>
              Trip Duration: undefined ms
            </p>
            <div
              class="wrapper"
            >
              <div
                class="info"
              >
                <div
                  class="image"
                >
                  <img
                    alt="Robert"
                    src="https://robohash.org/1234?size=50x50"
                  />
                </div>
                <div
                  class="car"
                >
                  <img
                    alt="car mode"
                    src="./images/van.svg"
                  />
                </div>
                <div>
                  <p
                    class="highlighted"
                  >
                    Herbert Patton
                    <span
                      class="rate"
                    >
                      <span
                        aria-label="rating"
                        role="img"
                      >
                        ‚≠êÔ∏è
                      </span>
                      5.0
                    </span>
                  </p>
                  <p>
                    Toyato HiAce - GA 2770
                  </p>
                </div>
              </div>
              <div
                class="destination"
              >
                <span>
                   New Cairo
                </span>
                <span>
                   Nasr City
                </span>
              </div>
              <dl
                class="about"
              >
                <div>
                  <dt>
                    <span
                      aria-label="location"
                      role="img"
                    >
                      üìç
                    </span>
                    Trip Distance:
                  </dt>
                  <dd>
                     KM
                  </dd>
                </div>
                <div>
                  <dt>
                    <span
                      aria-label="fare"
                      role="img"
                    >
                      üíµ
                    </span>
                    Trip Fare:
                  </dt>
                  <dd>
                    45 EGP
                  </dd>
                </div>
              </dl>
            </div>
          </section>
          <section>
            <h2>
              Bookings
            </h2>
            <ul
              class="list"
            />
          </section>
        </div>
      </div>
    </div>
    <div
      class="ReactModalPortal"
    />
  </body>,
  "container": <div>
    <div>
      <div
        style="height: 100%;"
      />
      <div
        class="container"
      >
        <div
          class="buttons"
        >
          <button
            class="button"
            name="Start ride"
            type="button"
          >
            Start ride
          </button>
          <button
            class="button"
            name="Book ride"
            type="button"
          >
            Book ride
          </button>
        </div>
        <section>
          <h2>
            Trip Information
          </h2>
          <p>
            Trip Duration: undefined ms
          </p>
          <div
            class="wrapper"
          >
            <div
              class="info"
            >
              <div
                class="image"
              >
                <img
                  alt="Robert"
                  src="https://robohash.org/1234?size=50x50"
                />
              </div>
              <div
                class="car"
              >
                <img
                  alt="car mode"
                  src="./images/van.svg"
                />
              </div>
              <div>
                <p
                  class="highlighted"
                >
                  Herbert Patton
                  <span
                    class="rate"
                  >
                    <span
                      aria-label="rating"
                      role="img"
                    >
                      ‚≠êÔ∏è
                    </span>
                    5.0
                  </span>
                </p>
                <p>
                  Toyato HiAce - GA 2770
                </p>
              </div>
            </div>
            <div
              class="destination"
            >
              <span>
                 New Cairo
              </span>
              <span>
                 Nasr City
              </span>
            </div>
            <dl
              class="about"
            >
              <div>
                <dt>
                  <span
                    aria-label="location"
                    role="img"
                  >
                    üìç
                  </span>
                  Trip Distance:
                </dt>
                <dd>
                   KM
                </dd>
              </div>
              <div>
                <dt>
                  <span
                    aria-label="fare"
                    role="img"
                  >
                    üíµ
                  </span>
                  Trip Fare:
                </dt>
                <dd>
                  45 EGP
                </dd>
              </div>
            </dl>
          </div>
        </section>
        <section>
          <h2>
            Bookings
          </h2>
          <ul
            class="list"
          />
        </section>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`snapshot trip 1 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          style="height: 100%;"
        />
        <div
          class="container"
        >
          <div
            class="buttons"
          >
            <button
              class="button"
              disabled=""
              name="Start ride"
              type="button"
            >
              Start ride
            </button>
            <button
              class="button"
              disabled=""
              name="Book ride"
              type="button"
            >
              Book ride
            </button>
          </div>
          <section>
            <h2>
              Trip Information
            </h2>
            <p>
              Trip Duration: undefined ms
            </p>
            <div
              class="wrapper"
            >
              <div
                class="info"
              >
                <div
                  class="image"
                >
                  <img
                    alt="Robert"
                    src="https://robohash.org/1234?size=50x50"
                  />
                </div>
                <div
                  class="car"
                >
                  <img
                    alt="car mode"
                    src="./images/van.svg"
                  />
                </div>
                <div>
                  <p
                    class="highlighted"
                  >
                    Herbert Patton
                    <span
                      class="rate"
                    >
                      <span
                        aria-label="rating"
                        role="img"
                      >
                        ‚≠êÔ∏è
                      </span>
                      5.0
                    </span>
                  </p>
                  <p>
                    Toyato HiAce - GA 2770
                  </p>
                </div>
              </div>
              <div
                class="destination"
              >
                <span>
                   New Cairo
                </span>
                <span>
                   Nasr City
                </span>
              </div>
              <dl
                class="about"
              >
                <div>
                  <dt>
                    <span
                      aria-label="location"
                      role="img"
                    >
                      üìç
                    </span>
                    Trip Distance:
                  </dt>
                  <dd>
                     KM
                  </dd>
                </div>
                <div>
                  <dt>
                    <span
                      aria-label="fare"
                      role="img"
                    >
                      üíµ
                    </span>
                    Trip Fare:
                  </dt>
                  <dd>
                    45 EGP
                  </dd>
                </div>
              </dl>
            </div>
          </section>
          <section>
            <h2>
              Bookings
            </h2>
            <ul
              class="list"
            />
          </section>
        </div>
      </div>
    </div>
    <div
      class="ReactModalPortal"
    />
  </body>,
  "container": <div>
    <div>
      <div
        style="height: 100%;"
      />
      <div
        class="container"
      >
        <div
          class="buttons"
        >
          <button
            class="button"
            disabled=""
            name="Start ride"
            type="button"
          >
            Start ride
          </button>
          <button
            class="button"
            disabled=""
            name="Book ride"
            type="button"
          >
            Book ride
          </button>
        </div>
        <section>
          <h2>
            Trip Information
          </h2>
          <p>
            Trip Duration: undefined ms
          </p>
          <div
            class="wrapper"
          >
            <div
              class="info"
            >
              <div
                class="image"
              >
                <img
                  alt="Robert"
                  src="https://robohash.org/1234?size=50x50"
                />
              </div>
              <div
                class="car"
              >
                <img
                  alt="car mode"
                  src="./images/van.svg"
                />
              </div>
              <div>
                <p
                  class="highlighted"
                >
                  Herbert Patton
                  <span
                    class="rate"
                  >
                    <span
                      aria-label="rating"
                      role="img"
                    >
                      ‚≠êÔ∏è
                    </span>
                    5.0
                  </span>
                </p>
                <p>
                  Toyato HiAce - GA 2770
                </p>
              </div>
            </div>
            <div
              class="destination"
            >
              <span>
                 New Cairo
              </span>
              <span>
                 Nasr City
              </span>
            </div>
            <dl
              class="about"
            >
              <div>
                <dt>
                  <span
                    aria-label="location"
                    role="img"
                  >
                    üìç
                  </span>
                  Trip Distance:
                </dt>
                <dd>
                   KM
                </dd>
              </div>
              <div>
                <dt>
                  <span
                    aria-label="fare"
                    role="img"
                  >
                    üíµ
                  </span>
                  Trip Fare:
                </dt>
                <dd>
                  45 EGP
                </dd>
              </div>
            </dl>
          </div>
        </section>
        <section>
          <h2>
            Bookings
          </h2>
          <ul
            class="list"
          />
        </section>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
